src/db/index.rs

```diff
@@ -187,80 +153,209 @@ fn get_index_keys(
     }])
 }
 
-// Returns the index key (index prolly map key) for an indexed value.
-pub fn index_key(secondary: IndexValue, primary: &[u8]) -> Result<Vec<u8>, GetIndexKeysError> {
+static KEY_VERSION_0: &[u8] = &[0];
+static KEY_SEPARATOR: &[u8] = &[0];
+
+// An index key is encoded to vec of bytes in the following order:
+//   - key version byte(s), followed by
+//   - the UTF8 encoded secondary key value, followed by
+//   - the key separator, a null byte, followed by
+//   - the primary key bytes
+//
+// The null separator byte ensures that if a secondary key A is longer than B then
+// A always sorts after B. Appending the primary key ensures index keys with
+// identical secondary keys sort in primary key order. Secondary keys must not
+// contain a null byte.
+pub fn encode_index_key(index_key: &IndexKey) -> Result<Vec<u8>, GetIndexKeysError> {
     use GetIndexKeysError::*;
 
-    let k = IndexKeyType::V0(IndexKey { secondary, primary });
-    Ok(bytekey::serialize(&k).map_err(|e| SerializeIndexEntryError(e.to_string()))?)
+    let IndexKey { secondary, primary } = *index_key;
+
+    if secondary.contains('\0') {
+        return Err(StringContainsNull(secondary.to_string()));
+    }
+    let mut v = Vec::new();
+    v.reserve(KEY_VERSION_0.len() + secondary.len() + KEY_SEPARATOR.len() + primary.len());
+    v.extend_from_slice(KEY_VERSION_0);
+    v.extend_from_slice(secondary.as_bytes());
+    v.extend_from_slice(KEY_SEPARATOR);
+    v.extend_from_slice(primary);
+    Ok(v)
 }
 
 // Returns bytes that can be used to scan for the given secondary index value. These bytes
-// are different from the full index key bytes in that we do *not* want the trailing null
-// byte that bytekey appends to the secondary index value ("key\0" is not a prefix of "keyfoo").
-pub fn scan_key(secondary: &str) -> Result<Vec<u8>, GetIndexKeysError> {
-    let mut k = index_key(IndexValue::Str(secondary), &[])?;
-    k.truncate(k.len() - 1);
+// are different from the full index key bytes in that we do *not* want the null
+// byte separating secondary and primary keys ("key\0" is not a prefix of "keyfoo").
+// Exclusive scans are implemented by appending a 0x01 byte to the end of the scan key,
+// making it start atthe next higher value.
+pub fn encode_scan_key(secondary: &str, exclusive: bool) -> Result<Vec<u8>, GetIndexKeysError> {
+    let mut k = encode_index_key(&IndexKey {
+        secondary,
+        primary: &[],
+    })?;
+    // Final byte will be the 0x00 separator. If exclusive, change it to 0x01 (the next
+    // value higher than secondary), otherwise strip it off.
+    k.pop();
+    if exclusive {
+        k.push(0x01);
+    }
     Ok(k)
 }
 
+// Decodes an IndexKey encoded by encode_index_key, returning the values as a tuple
+// since IndexKey holds references.
+// TODO use this to set key in scan results
+#[allow(dead_code)]
+pub fn decode_index_key(index_key: &[u8]) -> Result<(String, Vec<u8>), DecodeIndexKeyError> {
```

Comment:
Can the primary_bytes be a slice into the index key data?

Reply:
yes good point, done in https://github.com/rocicorp/repc/commit/56bfc8fe354d0d1477e8f653cc10ab3f30360c27